src/main/java/org/opensearch/knn/quantization/models/quantizationState/QuantizationStateCache.java

```diff
@@ -0,0 +1,132 @@
+            rebuildCache();
+        });
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(QUANTIZATION_STATE_CACHE_EXPIRY_TIME_MINUTES_SETTING, it -> {
+            rebuildCache();
+        });
+        buildCache();
+    }
+
+    /**
+     * Gets the singleton instance of the cache.
+     * @return QuantizationStateCache
+     */
+    public static QuantizationStateCache getInstance() {
+        if (instance == null) {
+            synchronized (QuantizationStateCache.class) {
+                if (instance == null) {
+                    instance = new QuantizationStateCache();
+                }
+            }
+        }
+        return instance;
+    }
+
+    /**
+     * Initialize the cache
+     * @param clusterService used to update settings
+     */
+    public static void initialize(ClusterService clusterService) {
+        QuantizationStateCache.clusterService = clusterService;
+    }
+
+    private void buildCache() {
+        this.cache = CacheBuilder.newBuilder().concurrencyLevel(1).maximumWeight(maxCacheSizeInKB).weigher((k, v) -> {
+            try {
+                return ((QuantizationState) v).toByteArray().length;
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        })
+            .expireAfterAccess(
+                QUANTIZATION_STATE_CACHE_EXPIRY_TIME_MINUTES_SETTING.get(clusterService.getSettings()).getMinutes(),
+                TimeUnit.MINUTES
+            )
+            .removalListener(this::onRemoval)
+            .build();
+    }
+
+    private void rebuildCache() {
+        clear();
+        buildCache();
+    }
+
+    /**
+     * Retrieves the quantization state associated with a given field name.
+     * @param fieldName The name of the field.
+     * @return The associated QuantizationState, or null if not present.
+     */
+    public QuantizationState getQuantizationState(String fieldName) {
+        return cache.getIfPresent(fieldName);
+    }
+
+    /**
+     * Adds or updates a quantization state in the cache.
+     * @param fieldName The name of the field.
+     * @param quantizationState The quantization state to store.
+     */
+    public void addQuantizationState(String fieldName, QuantizationState quantizationState) {
+        cache.put(fieldName, quantizationState);
+    }
+
+    /**
+     * Removes the quantization state associated with a given field name.
+     * @param fieldName The name of the field.
+     */
+    public void evict(String fieldName) {
+        cache.invalidate(fieldName);
+    }
+
+    private void onRemoval(RemovalNotification<String, QuantizationState> removalNotification) {
+        if (RemovalCause.SIZE == removalNotification.getCause()) {
+            updateEvictedDueToSizeAt();
+        }
+    }
+
+    private void updateEvictedDueToSizeAt() {
+        evictedDueToSizeAt = Instant.now();
+    }
+
+    /**
+     * Clears all entries from the cache.
+     */
+    public void clear() {
```

Comment:
any reason of making this function public?

Reply:
Eventually, it will be used when indices are closed, which separates it from the rebuild() method.